File: medio/conjunto/csharp/BinarySet.cs
using System;

public class BinarySet
{
    BinaryNode root;

    public void Add(int newElement) 
    {
        if (root == null) 
        {
            root = new BinaryNode(newElement);
            return;
        }

        BinaryNode node = root;

        while (node != null)
        {
            if (node.Element == newElement) return;
            
            if (newElement < node.Element) 
            {
                if (node.Less == null) 
                {
                    var nodeToAdd = new BinaryNode(newElement);
                    nodeToAdd.Parent = node;
                    nodeToAdd.Side = BinarySide.LEFT;
                    node.Less = nodeToAdd;
                    return;
                } 
                else 
                {
                    node = node.Less;
                }
            } 
            else 
            {
                if (node.Greater == null) 
                {
                    var nodeToAdd = new BinaryNode(newElement);
                    nodeToAdd.Parent = node;
                    nodeToAdd.Side = BinarySide.RIGHT;
                    node.Greater = nodeToAdd;
                    return;
                } 
                else 
                {
                    node = node.Greater;
                }
            } 
        }
    }

    public BinaryNode Find(int elementToSearch) 
    {
        BinaryNode node = root;

        while (node != null)
        {
            if (node.Element == elementToSearch) return node;
            node = elementToSearch < node.Element ? node.Less : node.Greater;
        }

        return null;
    }

    public bool Contains(int element) 
    {
        return Find(element) != null;
    }

    public int Count() 
    {
        if (root == null) return 0;
        return root.Count();
    }

    public void Remove(int element) 
    {
        BinaryNode nodeToRemove = Find(element);
        if (nodeToRemove == null) return;

        BinaryNode nodeToReplace = null;
        if (nodeToRemove.Less != null)  
        {
            nodeToReplace = nodeToRemove.Less.Max();
        } 
        else if (nodeToRemove.Greater != null) 
        {
            nodeToReplace = nodeToRemove.Greater.Min();
        }

        if (nodeToReplace != null) 
        {
            switch (nodeToReplace.Side)
            {
            case BinarySide.RIGHT:
                nodeToReplace.Parent.Greater = null;
                break;
            case BinarySide.LEFT:
                nodeToReplace.Parent.Less = null;
                break;
            }

            if (nodeToRemove.Less != null && nodeToRemove.Less.Element != nodeToReplace.Element) 
            {
                nodeToReplace.Less = nodeToRemove.Less;
                nodeToReplace.Less.Parent = nodeToReplace;
            }
            else 
            {
                nodeToReplace.Less = null;
            }

            if (nodeToRemove.Greater != null && nodeToRemove.Greater.Element != nodeToReplace.Element) 
            {
                nodeToReplace.Greater = nodeToRemove.Greater;
                nodeToReplace.Greater.Parent = nodeToReplace;
            }
            else 
            {
                nodeToReplace.Greater = null;
            }

            nodeToReplace.Parent = nodeToRemove.Parent;
            nodeToReplace.Side = nodeToRemove.Side;
        }

        switch (nodeToRemove.Side)
        {
        case BinarySide.RIGHT:
            nodeToRemove.Parent.Greater = nodeToReplace;
            break;
        case BinarySide.LEFT:
            nodeToRemove.Parent.Less = nodeToReplace;
            break;
        case BinarySide.NONE:
            root = nodeToReplace;
            break;
        }
    }
}

File: medio/conjunto/cpp/BinarySet.cpp
#include <iostream>
#include "BinarySet.h"

using namespace std;

void BinarySet::addElement(int newElement) 
{
    if (root == nullptr) 
    {
        root = new Node();
        root->element = newElement;
        root->side = NONE;
        root->greater = nullptr;
        root->less = nullptr;
    } 
    else 
    {
        auto node = root;
        while (node != nullptr) 
        {
            if (newElement == node->element) break;

            if (newElement < node->element) 
            {
                if (node->less == nullptr) 
                {
                    node->less = new Node();
                    node->less->element = newElement;
                    node->less->side = LEFT;
                    node->less->parent = node;
                    node->less->greater = nullptr;
                    node->less->less = nullptr;
                } 
                else 
                {
                    node = node->less;
                }
            } 
            else 
            {
                if (node->greater == nullptr) 
                {
                    node->greater = new Node();
                    node->greater->element = newElement;
                    node->greater->side = RIGHT;
                    node->greater->parent = node;
                    node->greater->greater = nullptr;
                    node->greater->less = nullptr;
                }
                else 
                {
                    node = node->greater;
                }
            }
        }          
    }
}

bool BinarySet::contains(int element) 
{
    return find(element);
}

Node* BinarySet::find(int element) 
{
    auto node = root;
    while (node != nullptr) 
    {
        if (node->element == element) 
        {
            return node;
        }
        node = (element < node->element) ? node->less : node->greater;
    }
    return nullptr;
}

Node* BinarySet::max(Node *node) 
{
    if (node == nullptr) return nullptr;
    while (node->greater != nullptr) 
    {
        node = node->greater;
    }
    return node;
}

Node* BinarySet::min(Node *node) 
{
    if (node == nullptr) return nullptr;
    while (node->less != nullptr) 
    {
        node = node->less;
    }
    return node;
}

void BinarySet::removeElement(int elementToRemove) 
{
    auto nodeToRemove = find(elementToRemove);
    
    if (nodeToRemove == nullptr) return;

    Node *toReplace = nullptr;
    
    if (nodeToRemove->less != nullptr)  
    {
        toReplace = max(nodeToRemove->less);
    } 
    else if (nodeToRemove->greater != nullptr) 
    {
        toReplace = min(nodeToRemove->greater);
    }
    
    if (toReplace != nullptr) 
    {
        switch (toReplace->side)
        {
        case RIGHT:
            toReplace->parent->greater = nullptr;
            break;
        case LEFT:
            toReplace->parent->less = nullptr;
            break;
        }

        if (nodeToRemove->less != nullptr && nodeToRemove->less->element != toReplace->element) 
        {
            toReplace->less = nodeToRemove->less;
            toReplace->less->parent = toReplace;
        }
        else 
        {
            toReplace->less = nullptr;
        }

        if (nodeToRemove->greater != nullptr && nodeToRemove->greater->element != toReplace->element) 
        {
            toReplace->greater = nodeToRemove->greater;
            toReplace->greater->parent = toReplace;
        }
        else 
        {
            toReplace->greater = nullptr;
        }

        toReplace->parent = nodeToRemove->parent;
        toReplace->side = nodeToRemove->side;
    }

    switch (nodeToRemove->side)
    {
    case RIGHT:
        nodeToRemove->parent->greater = toReplace;
        break;
    case LEFT:
        nodeToRemove->parent->less = toReplace;
        break;
    case NONE:
        root = toReplace;
        break;
    }

    delete nodeToRemove;
}

int BinarySet::count() {
    return count(root);
}

int BinarySet::count(Node* node) {
    int greater = node->greater == nullptr ? 0 : count(node->greater);
    int less = node->less == nullptr ? 0 : count(node->less);
    return 1 + greater + less;
}

void BinarySet::print() {
    print(root);
    cout << endl;
}

void BinarySet::print(Node* node) {
    cout << node->element << ":" << node->side << ", ";
    if (node->less != nullptr) print(node->less);
    if (node->greater != nullptr) print(node->greater);
}